router/core/websocket.go

```diff
@@ -158,76 +128,215 @@ func (c *wsConnectionWrapper) Close() error {
 type WebsocketHandler struct {
 	ctx              context.Context
 	next             http.Handler
-	ids              *globalIDStorage
 	parser           *OperationParser
 	planner          *OperationPlanner
 	graphqlHandler   *GraphQLHandler
 	metrics          *RouterMetrics
 	accessController *AccessController
 	logger           *zap.Logger
-}
 
-func isWsUpgradeRequest(r *http.Request) bool {
-	return r.Header.Get("Upgrade") == "websocket"
+	epoll         epoller.Poller
+	connections   map[int]*WebSocketConnectionHandler
+	connectionsMu sync.RWMutex
+
+	handlerPool   *pond.WorkerPool
+	connectionIDs atomic.Int64
+
+	stats *WebSocketStats
+
+	readTimeout time.Duration
 }
 
 func (h *WebsocketHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	// Don't call upgrader.Upgrade unless the request looks like a websocket
-	// because if Upgrade() fails it sends an error response
-	if isWsUpgradeRequest(r) {
-		// Check access control before upgrading the connection
-		validatedReq, err := h.accessController.Access(w, r)
-		if err != nil {
-			statusCode := http.StatusForbidden
-			if errors.Is(err, ErrUnauthorized) {
-				statusCode = http.StatusUnauthorized
-			}
-			http.Error(w, http.StatusText(statusCode), statusCode)
-			return
+	if !websocket.IsWebSocketUpgrade(r) {
+		h.next.ServeHTTP(w, r)
+		return
+	}
+	// Check access control before upgrading the connection
+	validatedReq, err := h.accessController.Access(w, r)
+	if err != nil {
+		statusCode := http.StatusForbidden
+		if errors.Is(err, ErrUnauthorized) {
+			statusCode = http.StatusUnauthorized
 		}
-		r = validatedReq
-
-		upgrader := websocket.Upgrader{
-			HandshakeTimeout: 5 * time.Second,
-			// TODO: WriteBufferPool,
-			EnableCompression: true,
-			Subprotocols:      wsproto.Subprotocols(),
-			CheckOrigin: func(_ *http.Request) bool {
-				// Allow any origin to subscribe via WS
+		http.Error(w, http.StatusText(statusCode), statusCode)
+		return
+	}
+	r = validatedReq
+	var (
+		subProtocol string
+	)
+	upgrader := ws.HTTPUpgrader{
```

Comment:
Can we reuse it across requests?

Reply:
yes, I think so but please test it

Reply:
I skipped this optimization because there is no easy way to access the protocol afterwards.